src/spark_rapids_ml/knn.py

```diff
@@ -0,0 +1,344 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union
+
+import cudf
+import numpy as np
+import pandas as pd
+from pyspark.ml.param.shared import HasInputCol, HasLabelCol, Param
+from pyspark.sql import Column, DataFrame
+from pyspark.sql.functions import col, lit
+from pyspark.sql.types import (
+    ArrayType,
+    DoubleType,
+    FloatType,
+    IntegerType,
+    Row,
+    StructField,
+    StructType,
+)
+
+from spark_rapids_ml.core import (
+    INIT_PARAMETERS_NAME,
+    CumlInputType,
+    CumlT,
+    _CumlEstimatorSupervised,
+    _CumlModel,
+    _set_pyspark_cuml_cls_param_attrs,
+    alias,
+)
+from spark_rapids_ml.params import _CumlClass, _CumlParams
+
+
+class NearestNeighborsClass(_CumlClass):
+    @classmethod
+    def _cuml_cls(cls) -> List[type]:
+        from cuml import NearestNeighbors as cumlNearestNeighbors
+
+        return [cumlNearestNeighbors]
+
+    @classmethod
+    def _param_mapping(cls) -> Dict[str, Optional[str]]:
+        return {}
+
+    @classmethod
+    def _param_excludes(cls) -> List[str]:
+        return [
+            "handle",
+            "algorithm",
+            "metric",
+            "p",
+            "algo_params",
+            "metric_expanded",
+            "metric_params",
+            "output_type",
+        ]
+
+
+class _NearestNeighborsCumlParams(_CumlParams, HasInputCol, HasLabelCol):
+    """
+    Shared Spark Params for NearestNeighbor and NearestNeighborModel.
+    """
+
+    def setInputCol(self, value: str) -> "_NearestNeighborsCumlParams":
+        """
+        Sets the value of `inputCol`.
+        """
+        self.set_params(inputCol=value)
+        return self
+
+
+class _CumlEstimatorNearestNeighbors(_CumlEstimatorSupervised):
+    """
+    Base class for Cuml Nearest Neighbor.
+    """
+
+    def _pre_process_data(
+        self, dataset: DataFrame
+    ) -> Tuple[
+        List[Column], Optional[List[str]], int, Union[Type[FloatType], Type[DoubleType]]
+    ]:
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = super()._pre_process_data(dataset)
+
+        select_cols.append(col(alias.row_number))
+
+        return select_cols, multi_col_names, dimension, feature_type
+
+
+class NearestNeighbors(
+    NearestNeighborsClass, _CumlEstimatorNearestNeighbors, _NearestNeighborsCumlParams
+):
+    """
+    Examples
+    --------
+    >>> from sparkcuml.neighbors import NearestNeighbors
+    >>> data = [([1.0, 1.0],),
+    ...         ([2.0, 2.0],),
+    ...         ([3.0, 3.0],),]
+    >>> topk = 2
+    >>> gpu_knn = NearestNeighbors().setInputCol("features").setK(topk)
+    >>> data_df = spark.SparkContext.parallelize(data).map(lambda row: (row,)).toDF(["features"])
```

Comment:
Is the map to tuple required here?  Doesn't seem required in unit tests.

Reply:
done.